# <span style="color:red">DATA CLEANING</span>

## <span style="color:blue">Check data</span>

In [3]:
df = raw_df.copy()
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

In [4]:
df = df.rename(columns={'age': 'Age', 
                          'job': 'Type of job',
                          'marital': 'Marital status',
                          'education': 'Education',
                          'default': 'Credit',
                          'housing': 'Housing loan',
                          'loan': 'Personal loan',
                          'contact': 'Contact communication type',
                          'month': 'Latest month',
                          'day_of_week': 'Latest day',
                          'duration': 'Latest duration',
                          'campaign': 'Number of contacts',
                          'pdays': 'Passed days',
                          'previous': "Previous campaign's number",
                          'poutcome': "Previous campaign outcome",
                          'y': 'Label'})

In [5]:
df['Label'] = df['Label'].replace({'yes': 1, 'no': 0, 'unknown': pd.NA}).astype('Int64')
df = df.replace('unknown', pd.NA)

# The value 999 in the 'Passed days' column was recoded to -1 to denote missing contact history. An additional binary column was created to flag customers who were never contacted
df['Passed days'] = df['Passed days'].replace(999, pd.NA)
def contactstatus(x):
    if pd.isna(x):
        return 'Never contacted'
    else:
        return 'Was previously contacted'
df['Contact status'] = df['Passed days'].apply(contactstatus)
df['Passed days'] = df['Passed days'].fillna(-1).astype(int)

def Agerange(x):
    if x < 21:
        return '16-20'
    elif x < 26:
        return '21-25'
    elif x < 31:
        return '26-30'
    elif x < 36:
        return '31-35'
    elif x < 41:
        return '36-40'
    elif x < 46:
        return '41-45'
    elif x < 51:
        return '46-50'
    elif x < 56:
        return '51-55'
    elif x < 61:
        return '56-60'
    elif x < 66:
        return '61-65'
    elif x < 71:
        return '66-70'
    elif x < 76:
        return '71-75'
    elif x < 81:
        return '76-80'
    elif x < 86:
        return '81-85'
    elif x < 91:
        return '86-90'
    elif x < 96:
        return '91-95'
    elif x < 101:
        return '96-100'

df['Age range'] = df['Age'].apply(Agerange)

def Durationrange(x):
    if x == 0:
        return '00min'
    elif x < 301:
        return '<06mins'
    elif x < 601:
        return '06-10mins'
    elif x < 901:
        return '11-15mins'
    elif x < 1201:
        return '16-20mins'
    elif x < 1501:
        return '21-25mins'
    elif x < 1801:
        return '26-30mins'
    elif x < 2101:
        return '31-35mins'
    elif x < 2401:
        return '36-40mins'
    elif x < 2701:
        return '41-45mins'
    elif x < 3001:
        return '46-50mins'
    elif x < 3301:
        return '51-55mins'
    elif x < 3601:
        return '56-60mins'
    elif x < 3901:
        return '61-65mins'
    elif x < 4201:
        return '66-70mins'
    elif x < 4501:
        return '71-75mins'
    elif x < 4801:
        return '76-80mins'
    elif x < 5101:
        return '81-85mins'

df['Latest duration range'] = df['Latest duration'].apply(Durationrange)

In [6]:
df.duplicated().sum()

np.int64(12)

In [7]:
df.isna().sum()

Age                              0
Type of job                    330
Marital status                  80
Education                     1731
Credit                        8597
Housing loan                   990
Personal loan                  990
Contact communication type       0
Latest month                     0
Latest day                       0
Latest duration                  0
Number of contacts               0
Passed days                      0
Previous campaign's number       0
Previous campaign outcome        0
emp.var.rate                     0
cons.price.idx                   0
cons.conf.idx                    0
euribor3m                        0
nr.employed                      0
Label                            0
Contact status                   0
Age range                        0
Latest duration range            0
dtype: int64

## <span style="color:blue">Clean data</span>

##### **The duplicate cases**
These cases are identical in the columns: duration, last contact day, last contact month, emp.var.rate, cons.price.idx, cons.conf.idx, euribor3m, nr.employed  
→ These are unlikely to occur naturally, so they can be safely removed 

In [8]:
df[df.duplicated(keep=False)]

Age  Type of job Marital status            Education Credit  \
1265    39  blue-collar        married             basic.6y     no   
1266    39  blue-collar        married             basic.6y     no   
12260   36      retired        married                 <NA>     no   
12261   36      retired        married                 <NA>     no   
14155   27   technician         single  professional.course     no   
14234   27   technician         single  professional.course     no   
16819   47   technician       divorced          high.school     no   
16956   47   technician       divorced          high.school     no   
18464   32   technician         single  professional.course     no   
18465   32   technician         single  professional.course     no   
20072   55     services        married          high.school   <NA>   
20216   55     services        married          high.school   <NA>   
20531   41   technician        married  professional.course     no   
20534   41   technician        married  professional.course     no   
25183   39       admin.        married    university.degree     no   
25217   39       admin.        married    university.degree     no   
28476   24     services         single          high.school     no   
28477   24     services         single          high.school     no   
32505   35       admin.        married    university.degree     no   
32516   35       admin.        married    university.degree     no   
36950   45       admin.        married    university.degree     no   
36951   45       admin.        married    university.degree     no   
38255   71      retired         single    university.degree     no   
38281   71      retired         single    university.degree     no   

      Housing loan Personal loan Contact communication type Latest month  \
1265            no            no                  telephone          may   
1266            no            no                  telephone          may   
12260           no            no                  telephone          jul   
12261           no            no                  telephone          jul   
14155           no            no                   cellular          jul   
14234           no            no                   cellular          jul   
16819          yes            no                   cellular          jul   
16956          yes            no                   cellular          jul   
18464          yes            no                   cellular          jul   
18465          yes            no                   cellular          jul   
20072           no            no                   cellular          aug   
20216           no            no                   cellular          aug   
20531          yes            no                   cellular          aug   
20534          yes            no                   cellular          aug   
25183           no            no                   cellular          nov   
25217           no            no                   cellular          nov   
28476          yes            no                   cellular          apr   
28477          yes            no                   cellular          apr   
32505          yes            no                   cellular          may   
32516          yes            no                   cellular          may   
36950           no            no                   cellular          jul   
36951           no            no                   cellular          jul   
38255           no            no                  telephone          oct   
38281           no            no                  telephone          oct   

      Latest day  ...  Previous campaign outcome  emp.var.rate  \
1265         thu  ...                nonexistent           1.1   
1266         thu  ...                nonexistent           1.1   
12260        thu  ...                nonexistent           1.4   
12261        thu  ...                nonexistent           1.4   
14155        mon  ...    

In [9]:
df.drop_duplicates(inplace=True)

##### **Missing values rows** 
These rows in the dataset need to be considered for removal to ensure accuracy and efficiency in analysis:
- Type of job: 330 missing values, object
- Marital status: 80 missing values, object
- Education: 1731 missing values, object
- Housing loan: 990 missing values, boolen
- Personal loan: 990 missing values, boolen

In [10]:
df.dropna(subset=["Type of job", 'Marital status', 'Education', 'Housing loan', 'Personal loan'], inplace=True)

##### **Missing values columns** 
These columns in the dataset need to be considered for removal to ensure accuracy and efficiency in analysis:
- Contact communication type: It is not essential to distinguishes between 'cellular' and 'telephone'. Many people primarily use their mobile phones for all calls, whether to other mobile numbers or landlines. From a contact perspective, both essentially represent a phone number for direct communication.
- Credit: The 'Credit' column has 8,597 missing values, with only 3 non-missing entries indicating actual credit usage. The feature is not expected to hold significant analytical value. Consequently, it is recommended to remove this column from the dataset.

In [11]:
df['Contact communication type'].value_counts()

Contact communication type
cellular     24432
telephone    13802
Name: count, dtype: int64

In [12]:
df_credit1 = df[df['Label'] == 0].groupby('Credit')['Label'].count()
df_credit2 = df[df['Label'] == 1].groupby('Credit')['Label'].count()
df_credit = pd.merge(df_credit1, df_credit2, on='Credit', how='outer').reset_index()
df_credit.rename(columns={"Label_x": "Label Yes", "Label_y": "Label No"}, inplace=True)
df_credit

Credit  Label Yes  Label No
0     no      26617      3858
1    yes          3      <NA>

In [13]:
df.drop(columns=['Contact communication type', 'Credit'], inplace=True)

##### **The macroeconomic factor**
These columns are not included in the analysis:
- Emp.var.rate: employment variation rate - quarterly indicator (numeric)
- Cons.price.idx: consumer price index - monthly indicator (numeric)
- Cons.conf.idx: consumer confidence index - monthly indicator (numeric)
- Euribor3m: euribor 3 month rate - daily indicator (numeric)
- Nr.employed: number of employees - quarterly indicator (numeric)

In [14]:
df.drop(columns=['emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed'], inplace=True)

## <span style="color:blue">Check outlier</span>

In [15]:
df.describe()

Age  Latest duration  Number of contacts   Passed days  \
count  38234.000000     38234.000000        38234.000000  38234.000000   
mean      39.860438       258.236282            2.566930     -0.751190   
std       10.288704       259.818894            2.767777      1.478077   
min       17.000000         0.000000            1.000000     -1.000000   
25%       32.000000       102.000000            1.000000     -1.000000   
50%       38.000000       180.000000            2.000000     -1.000000   
75%       47.000000       319.000000            3.000000     -1.000000   
max       98.000000      4918.000000           43.000000     27.000000   

       Previous campaign's number     Label  
count                38234.000000   38234.0  
mean                     0.170058  0.111341  
std                      0.487231  0.314558  
min                      0.000000       0.0  
25%                      0.000000       0.0  
50%                      0.000000       0.0  
75%                      0.000000       0.0  
max                      7.000000       1.0

**Lastest Call Duration**  
→ Some calls last longer than usual, but since they involve financial telesales, these durations are reasonable — so we kept them in the dataset.

In [16]:
def outlier(column_name, df):
    q1 = df[column_name].quantile(0.25)
    q3 = df[column_name].quantile(0.75)
    iqr = q3 - q1
    upper = q3 + 1.5 * iqr
    lower = q1 - 1.5 * iqr
    lower_outliers = df[df[column_name] < lower][column_name].count()
    upper_outliers = df[df[column_name] > upper][column_name].count()
    print(f"upper = {upper}, lower = {lower}")
    print(f"Lower outlier count: {lower_outliers}, %Lower outlier: {lower_outliers/df[column_name].count() * 100:.2f}%")
    print(f"Upper outlier count: {upper_outliers}, %Upper outlier: {upper_outliers/df[column_name].count() * 100:.2f}%")

outlier("Latest duration", df)

upper = 644.5, lower = -223.5
Lower outlier count: 0, %Lower outlier: 0.00%
Upper outlier count: 2749, %Upper outlier: 7.19%
